File: github_code_index_sort/cpp/4/84.c
/* C++ Program to Implement Heap Sort
 * Ref: Introduction to Algorithms,CLRS Page:150
 * KUMAR GAURAV
 * DEC20,2014*/
#include <bits/stdc++.h>
using namespace std;

struct Maxheap{
	int *arr;
	int heapsize;
	};
void swap(int *a,int*b)
{
	int t;
	t=*a;
	*a=*b;
	*b=t;
}
void Maxheapify(struct Maxheap* m,int i)
{
	int largest=i;
	int left=(i<<1)+1;
	int right=(i+1)<<1;													//chosing the index of the largest element 
	if(left < m->heapsize && m->arr[left] > m->arr[largest])
		largest=left;
	if(right < m->heapsize && m->arr[right] > m->arr[largest])
			largest=right;
	if(largest!=i){														//setting largest as root
		swap(&m->arr[largest],&m->arr[i]);								//then calling Maxheapify to fix the heap from the child
		Maxheapify(m,largest);
	}
}

struct Maxheap*  BuildMaxHeap(int *a,int n)
{
	struct Maxheap* m = (struct Maxheap*)malloc(sizeof(struct Maxheap));
	m->arr	= a;
	m->heapsize=n;
	for(int i=(m->heapsize-2)/2;i>=0;i--)
	{
		Maxheapify(m,i);
	}
	return m;
}

void heapsort(int a[],int n)											
{		
	struct Maxheap* m = BuildMaxHeap(a,n);								//Building the heap using the input array and size
	for(int i=m->heapsize-1;i>=1;i--)									//one by one copying the root to the last position and fixing the heap each time
	{																	//by calling Maxheapify() 		
		swap(&m->arr[0],&m->arr[i]);
		m->heapsize--;
		Maxheapify(m,0);
	}
}

int main()
{
	int a[100],i,n;
	cout<<"Enter Size \t";
	cin>>n;
	cout<<"\nEnter The array Elements\n";
	for(i=0;i<n;i++)
	{
		cin>>a[i];
	}
	heapsort(a,n);														//calling the sorting function	
	cout<<"The Sorted Array is : \n";
	for(i=0;i<n;i++)
	{
		cout<<a[i]<<" ";
	}
	
}

File: github_java_sort/5/84.java
package com.mpakhomov.algorithms.sort;


public class BubbleSort implements Sorting {

    
    static public void sortNaive(int[] a) {
        final int N = a.length;
        for (int i = 0; i < N; i++) {
            for (int j = 1; j < N - i; j++) {
                if (a[j - 1] > a[j]) {
                    swap(a, j - 1, j);
                }
            }
        }
    }

    
    static public void sort(int[] a) {
        boolean swapped = true;
        for (int length = a.length; swapped ; length--) {
            swapped = false;
            for (int j = 1; j < length; j++) {
                if (a[j - 1] > a[j]) {
                    swap(a, j - 1, j);
                    swapped = true;
                }
            }
        }
    }

    static private void swap(int[] a, int i, int j) {
        int tmp = a[i];
        a[i] = a[j];
        a[j] = tmp;
    }

    @Override
    public void instanceSort(int[] a) {
        sort(a);
    }
}
